Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -30,167 +30,320 @@
 #import "LoggingNative.h"
 #import "NSOperationInternal.h"
 
+#include <deque>
+#include <mutex>
+#include <condition_variable>
+#include <algorithm>
+#include <memory>
+
 static const wchar_t* TAG = L"NSOperationQueue";
 
-typedef void* gpointer;
+class NSOperationQueuePriv {
+public:
+    NSOperationQueuePriv(NSOperationQueue* parent) : _thread(nullptr), _maxConcurrentOperationCount(1), _isSuspended(NO) {
+        _name.attach([[NSString alloc] initWithFormat:@"NSOperationQueue %08x", parent]);
+    }
 
-struct NSAtomicListNode {
-    struct NSAtomicListNode* next;
-    StrongId<NSObject> elt;
-};
-typedef struct NSAtomicListNode* NSAtomicListRef;
+    NSString* Name() {
+        std::lock_guard<std::mutex> lock(_lock);
+        return _name;
+    }
 
-struct NSOperationQueuePriv {
-    NSAtomicListRef myQueues[NSOperationQueuePriority_Count];
-    NSAtomicListRef queues[NSOperationQueuePriority_Count];
+    void SetName(NSString* name) {
+        std::lock_guard<std::mutex> lock(_lock);
+        _name.attach([name copy]);
+    }
 
-    NSThread* _thread;
-    pthread_mutex_t _threadRunningLock;
+    BOOL RunOperationFromLists(unsigned int priority) {
+        StrongId<NSOperation> candidateOperation;
+
+        {
+            std::lock_guard<std::mutex> lock(_lock);
+            if (_candidateOperationsToRunNext[priority].empty()) {
+                // No operations "on deck". Go pull some more in from the holding area.
+                // Holding area lists are in LIFO order, but we want to execute operations in the order they were enqueued
+                // so we reverse the list before we do anything with it.
+                _candidateOperationsToRunNext[priority].swap(_operationsHoldingArea[priority]);
+                std::reverse(_candidateOperationsToRunNext[priority].begin(), _candidateOperationsToRunNext[priority].end());
+            }
 
-    DWORD _maxConcurrentOperationCount;
-    id workAvailable;
-    id suspendedCondition;
-    id allWorkDone;
-    id curOperation;
-    BOOL isSuspended;
+            if (!_candidateOperationsToRunNext[priority].empty()) {
+                candidateOperation = _candidateOperationsToRunNext[priority].back();
+                _candidateOperationsToRunNext[priority].pop_back();
+            }
+        }
 
-    id _name;
+        // Now that a candidate has been determined, drop the lock so that it can potientially run.
+        // NOTE: a [operation start] call will potienitally run the operation on that this thread.
```

Comment:
this.

---
In reply to: [95449581](https://github.com/Microsoft/WinObjC/pull/1676#discussion_r95449581) [](ancestors = 95449581)